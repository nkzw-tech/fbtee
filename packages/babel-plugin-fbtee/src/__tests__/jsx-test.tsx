import { describe, expect, it, test } from '@jest/globals';
import {
  jsCodeFbtCallSerializer,
  snapshotTransformKeepJsx,
  withFbtImportStatement,
} from './FbtTestUtil.tsx';

expect.addSnapshotSerializer(jsCodeFbtCallSerializer);

describe('Test declarative (jsx) fbt syntax translation', () => {
  test('Enable explicit whitespace', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="squelched">
            <fbt:param name="one">{one}</fbt:param>
            {" "}
            <fbt:param name="two">{two}</fbt:param>
            {\` \`}
            <fbt:param name="three">{three}</fbt:param>
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('Squelch whitespace when in an expression', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x =
            <fbt desc="squelched">
              {"Squelched white space... "}
              with some
              {' other stuff.'}
            </fbt>;
            baz();`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('fbt:param with multiple children should error', () => {
    expect(() =>
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="some-desc">
            <fbt:param name="foo">
              {foo}
              {bar}
            </fbt:param>
          </fbt>`,
        ),
      ),
    ).toThrow(`fbt:param expects an {expression} or JSX element, and only one`);
  });

  test('fbt:param with multiple empty expression containers should be ok', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="some-desc">
            <fbt:param name="foo">
              {}
              {/* comment */}
              {foo}
              {}
            </fbt:param>
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should be able to house arbitrary markup within fbt:param nodes', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<div>
            <fbt desc="...">
              <fbt:param name="time">{formatDate(date, "F d, Y")}</fbt:param>
                by
              <fbt:param name="user name">
                <Link href={{url:user.link}}>
                  {user.name}
                </Link>
              </fbt:param>
            </fbt>
          </div>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should be able to nest within React nodes', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <div>
              <fbt desc="nested!">
                A nested string
              </fbt>
            </div>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should convert simple strings', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="It's simple">A simple string</fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should correctly destruct expression values in options', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="d">str
            <fbt:param name="count" number={someNum}>
              {getNum()}
            </fbt:param>
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should filter comment and empty expressions from children', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="It's simple">
            {
            }A sim{/*
              ignore
              me
              */}ple s{ }tri{}ng{/*ignore me*/}</fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle common string', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`<fbt common={true}>Done</fbt>`),
        { fbtCommon: { Done: 'The description for the common string "Done"' } },
      ),
    ).toMatchSnapshot();
  });

  test('should handle concatenated descriptions', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc={"A very long description " + "that we will concatenate " +
            "a few times"}
            project={"With" + "a" + "project"}>
            Here it is
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle empty string', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="a message!">
            A parameterized message to:
            <fbt:param name="emptyString"> </fbt:param>
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle enums (with array values)', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="enums!">
            Click to see
            <fbt:enum
              enum-range={[
                "groups",
                "photos",
                "videos"
              ]}
              value={id}
            />
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle enums (with object values)', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="enums!">
              Click to see
              <fbt:enum
                enum-range={{
                  id1: "groups",
                  id2: "photos",
                  id3: "videos"
                }}
                value={id}
              />
            </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle enums with more text', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="enums!">
            Click to see
            <fbt:enum
              enum-range={{
                id1: "groups",
                id2: "photos",
                id3: "videos"
              }}
              value={id}
            />
            Hey-hey!
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle fbt common attribute without value', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`<fbt common>Okay</fbt>`),
        { fbtCommon: { Okay: 'The description for the common string "Okay"' } },
      ),
    ).toMatchSnapshot();
  });

  test('should handle number={true} - (same output as above test)', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="variations!">
            Click to see
            <fbt:param name="count" number={true}>{c}</fbt:param>
            links
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle object pronoun', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc={"d"} project={"p"}>
            I know <fbt:pronoun type="object" gender={gender}/>.
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle params', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="a message!">
              A parameterized message to:
              <fbt:param name="personName">{theName}</fbt:param>
            </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle single expression with concatenated strings', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="foo">
            {"foo" + "bar"}
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle subject+reflexive pronouns', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc={"d"} project={"p"}>
            <fbt:pronoun type="subject" gender={gender} capitalize={true} human={true}/>
            wished <fbt:pronoun type="reflexive" gender={gender} human={true}/> a happy birthday.
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle template descriptions', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc={\`A very long description
            that will be a
            template across multiple lines\`}
            project={"With" + "a" + "project"}>
            Here it is
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should handle variations', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x = <fbt desc="variations!">
            Click to see
            <fbt:param name="count" number="true">{c}</fbt:param>
            links
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should ignore __private attributes', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt __self="fbt" desc="some-desc">
            <fbt:param __self="param" name="foo">
              {foo}
            </fbt:param>
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should ignore non-expression children in fbt:param', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="some-desc">
            <fbt:param name="foo">
              !{foo}!
            </fbt:param>
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should insert param value for same-param', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="d">str
            <fbt:param name="foo">{Bar}</fbt:param> and
            <fbt:same-param name="foo"/>
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should maintain order of params and enums', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="some-desc">
            Hello,
            <fbt:param name="foo">
              {foo}
            </fbt:param>
            <fbt:enum enum-range={["x", "y"]} value={x} />
            <fbt:param name="bar" number={n}>
              {bar}
            </fbt:param>
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should not insert extra space', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="Greating in i18n demo">
            Hello, <fbt:param name="guest">
              {guest}
            </fbt:param>!
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should strip out more newlines', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x =
            <fbt desc="moar lines">
              A simple string...
              with some other stuff.
            </fbt>;
            baz();`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should strip out newlines', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `var x =
            <fbt desc="Test trailing space when not last child">
              Preamble
              <fbt:param name="parm">{blah}</fbt:param>
            </fbt>;
          baz();`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should support html escapes', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`<fbt desc="foo &quot;bar&quot;">&times;</fbt>`),
      ),
    ).toMatchSnapshot();
  });

  test('should support non-breaking space character', () => {
    // multiple spaces are normalized to a single space
    // but &nbsp characters are preserved
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="desc with    non-breaking&nbsp;&nbsp;&nbsp;space">
              text with    non-breaking&nbsp;&nbsp;&nbsp;space
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should support unicode characters', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `// A backslash \\ in comments
          <fbt desc="unicode characters">
            A copyright sign {'\\u00A9'},
            a multi byte character {'\\uD83D\\uDCA9'},
            and a backslash {'\\\\'}.
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should throw for fbt that has description and common attribute (without value)', () => {
    expect(() =>
      snapshotTransformKeepJsx(
        withFbtImportStatement(`<fbt common={true} desc='d'>No</fbt>`),
        { fbtCommon: { No: 'The description for the common string "No"' } },
      ),
    ).toThrow(`<fbt common> must not have "desc" attribute.`);
  });

  test('should throw for strings with `common` attribute equal to false', () => {
    expect(() =>
      snapshotTransformKeepJsx(
        withFbtImportStatement(`<fbt common={false}>Yes</fbt>`),
        { fbtCommon: { Yes: 'The description for the common string "Yes"' } },
      ),
    ).toThrow(`This node requires a 'desc' attribute.`);
  });

  test('should throw on invalid attributes in fbt:param', () => {
    expect(() =>
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="some-desc">
            <fbt:param name="foo" qux="foo" desc="foo-desc">
              {foo}
            </fbt:param>
          </fbt>`,
        ),
      ),
    ).toThrow(`Invalid option "qux". Only allowed: gender, number, name`);
  });

  test('should throw on undefined common string', () => {
    expect(() =>
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt common={true}>Some undefined common string</fbt>`,
        ),
        {},
      ),
    ).toThrow(
      `Unknown string "Some undefined common string" for <fbt common={true}>`,
    );
  });

  test('should treat multiline descs as a single line', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<fbt desc="hi how are you today im doing well i guess
            how is your mother is she well yeah why not lets go
            home and never come back.">
            lol
          </fbt>`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should work with fragments', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<Fragment>
            <fbt desc="...">
              <fbt:param name="time">{formatDate(date, "F d, Y")}</fbt:param>
               by
              <fbt:param name="user name">
                <Link href={{url:user.link}}>
                  {user.name}
                </Link>
              </fbt:param>
            </fbt>
          </Fragment>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  test('should work with implicit fragments', () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(
          `<>
            <fbt desc="...">
              <fbt:param name="time">{formatDate(date, "F d, Y")}</fbt:param>
               by
              <fbt:param name="user name">
                <Link href={{url:user.link}}>
                  {user.name}
                </Link>
              </fbt:param>
            </fbt>
          </>;`,
        ),
      ),
    ).toMatchSnapshot();
  });
});

describe('Test fbt transforms without the jsx transform', () => {
  it('not nested', async () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`
        let x =
          <fbt desc="nested!">
            A nested string
          </fbt>;
      `),
      ),
    ).toMatchSnapshot(); // Should be like fbt._()
  });

  it('nested in div', async () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`
        let x =
          <div>
            <fbt desc="nested!">
              A nested string
            </fbt>
          </div>;
      `),
      ),
    ).toMatchSnapshot(); // Should be like <div>{fbt._()}</div>
  });

  it('short bool syntax for doNotExtract attribute', async () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`
        let x = <fbt desc="" doNotExtract>Test</fbt>;
      `),
      ),
    ).toMatchSnapshot();
  });

  it('short bool syntax for number attribute', async () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`
        let x =
          <fbt desc="">
            <fbt:param name="name" number>{'name'}</fbt:param>
          </fbt>;
      `),
      ),
    ).toMatchSnapshot();
  });

  describe('when using within template literals', () => {
    it('should work with a basic <fbt>', async () => {
      expect(
        snapshotTransformKeepJsx(
          withFbtImportStatement(`
          html\`<div>
            \${
              <fbt desc="some desc" project="some project">
                basic text
              </fbt>
            }
          </div>\`;
        `),
        ),
      ).toMatchSnapshot();
    });

    it('should work with basic <fbt> auto-parameterization', async () => {
      expect(
        snapshotTransformKeepJsx(
          withFbtImportStatement(`
          html\`<div>
            \${
              <fbt desc="some desc" project="some project">
                outer text
                <strong>
                  bold text
                </strong>
              </fbt>
            }
          </div>\`;
        `),
        ),
      ).toMatchSnapshot();
    });

    it('should dedupe plurals', async () => {
      expect(
        snapshotTransformKeepJsx(
          withFbtImportStatement(`
          <fbt desc="desc...">
            There
            <fbt:plural count={num} many="are">is</fbt:plural>{' '}
            <fbt:plural count={num} showCount="yes" value={hi()}>
              photo
            </fbt:plural>.
          </fbt>
        `),
        ),
      ).toMatchSnapshot();
    });

    it('should work with a nested <fbt> within an <fbt:param>', async () => {
      expect(
        snapshotTransformKeepJsx(
          withFbtImportStatement(`
          html\`<div>
            \${
              <fbt desc="some desc" project="some project">
                outer text
                <fbt:param name="param text">
                  {
                    html\`<strong>
                      \${
                        <fbt desc="inner string">
                          inner text
                          <fbt:param name="inner param">
                            {'bold'}
                          </fbt:param>
                        </fbt>
                      }
                    </strong>\`
                  }
                </fbt:param>
              </fbt>
            }
          </div>\`;
        `),
        ),
      ).toMatchSnapshot();
    });
  });

  // TODO: Fix preserving whitespace for JSX text.
  it('should fail to preserve whitespace in text when preserveWhitespace=true (known bug)', async () => {
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`
        <fbt desc="desc with 3   spaces" preserveWhitespace={true}>
          Some text with 3   spaces in between.
        </fbt>;
      `),
      ),
    ).toMatchSnapshot();
  });

  // TODO: We should NOT insert a space between two <fbt:plural>'s
  it(`[legacy buggy behavior] <fbt:pronoun> should insert a space character between two fbt constructs that don't neighbor raw text`, async () =>
    expect(
      snapshotTransformKeepJsx(
        withFbtImportStatement(`
        <fbt desc="">
          You can add
          <fbt:plural count={count} many="these">
            this
          </fbt:plural>
          <fbt:plural count={count} many="tags">
            tag
          </fbt:plural>
          to anything.
        </fbt>
      `),
      ),
    ).toMatchSnapshot());
});

test('Test common fbt with value-less `common` attribute should have same runtime call as the regular common fbt', async () => {
  const options = {
    fbtCommon: { Submit: 'The description for the common string "Submit"' },
  };
  expect(
    snapshotTransformKeepJsx(
      withFbtImportStatement(`
        let x = <fbt common>Submit</fbt>;
      `),
      options,
    ),
  ).toEqual(
    snapshotTransformKeepJsx(
      withFbtImportStatement(`
        let x = <fbt common={true}>Submit</fbt>;
      `),
      options,
    ),
  );
});
