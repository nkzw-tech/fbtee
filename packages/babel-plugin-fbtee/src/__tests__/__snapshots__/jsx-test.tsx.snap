// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test declarative (jsx) fbt syntax translation Enable explicit whitespace 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  "{one} {two} {three}",
  [fbt._param("one", one), fbt._param("two", two), fbt._param("three", three)],
  {
    hk: "1AF7e1",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation Squelch whitespace when in an expression 1`] = `
import { fbt } from "fbtee";
var x = fbt._("Squelched white space... with some other stuff.", null, {
  hk: "UtpyE",
});
baz();

`;

exports[`Test declarative (jsx) fbt syntax translation fbt:param with multiple empty expression containers should be ok 1`] = `
import { fbt } from "fbtee";
fbt._("{foo}", [fbt._param("foo", foo)], {
  hk: "2kCEE7",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should be able to house arbitrary markup within fbt:param nodes 1`] = `
import { fbt } from "fbtee";
<div>
  {fbt._(
    "{time} by {user name}",
    [
      fbt._param("time", formatDate(date, "F d, Y")),
      fbt._param(
        "user name",
        <Link
          href={{
            url: user.link,
          }}
        >
          {user.name}
        </Link>
      ),
    ],
    {
      hk: "1KGso4",
    }
  )}
</div>;

`;

exports[`Test declarative (jsx) fbt syntax translation should be able to nest within React nodes 1`] = `
import { fbt } from "fbtee";
var x = (
  <div>
    {fbt._("A nested string", null, {
      hk: "2Z1yyy",
    })}
  </div>
);

`;

exports[`Test declarative (jsx) fbt syntax translation should convert simple strings 1`] = `
import { fbt } from "fbtee";
var x = fbt._("A simple string", null, {
  hk: "pITkM",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should correctly destruct expression values in options 1`] = `
import { fbt } from "fbtee";
fbt._(
  {
    "*": "str {count}",
  },
  [fbt._param("count", getNum(), [0, someNum])],
  {
    hk: "A8zRg",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should filter comment and empty expressions from children 1`] = `
import { fbt } from "fbtee";
var x = fbt._("A simple string", null, {
  hk: "pITkM",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should handle common string 1`] = `
import { fbt } from "fbtee";
fbt._("Done", null, {
  hk: "3wsyxm",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should handle concatenated descriptions 1`] = `
import { fbt } from "fbtee";
fbt._("Here it is", null, {
  hk: "bwmzh",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should handle empty string 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  "A parameterized message to: {emptyString}",
  [fbt._param("emptyString", " ")],
  {
    hk: "1k5iwm",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle enums (with array values) 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  {
    groups: "Click to see groups",
    photos: "Click to see photos",
    videos: "Click to see videos",
  },
  [
    fbt._enum(id, {
      groups: "groups",
      photos: "photos",
      videos: "videos",
    }),
  ],
  {
    hk: "2bHIKd",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle enums (with object values) 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  {
    id1: "Click to see groups",
    id2: "Click to see photos",
    id3: "Click to see videos",
  },
  [
    fbt._enum(id, {
      id1: "groups",
      id2: "photos",
      id3: "videos",
    }),
  ],
  {
    hk: "3SHnwE",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle enums with more text 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  {
    id1: "Click to see groups Hey-hey!",
    id2: "Click to see photos Hey-hey!",
    id3: "Click to see videos Hey-hey!",
  },
  [
    fbt._enum(id, {
      id1: "groups",
      id2: "photos",
      id3: "videos",
    }),
  ],
  {
    hk: "328Iya",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle fbt common attribute without value 1`] = `
import { fbt } from "fbtee";
fbt._("Okay", null, {
  hk: "2aLqR1",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should handle number={true} - (same output as above test) 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  {
    "*": "Click to see {count} links",
  },
  [fbt._param("count", c, [0])],
  {
    hk: "35KbmT",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle object pronoun 1`] = `
import { fbt } from "fbtee";
fbt._(
  {
    0: "I know this.",
    1: "I know her.",
    2: "I know him.",
    "*": "I know them.",
  },
  [fbt._pronoun(0, gender)],
  {
    hk: "19l1pH",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle params 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  "A parameterized message to: {personName}",
  [fbt._param("personName", theName)],
  {
    hk: "2ArYIU",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle single expression with concentated strings 1`] = `
import { fbt } from "fbtee";
fbt._("foobar", null, {
  hk: "4l7JLw",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should handle subject+reflexive pronouns 1`] = `
import { fbt } from "fbtee";
fbt._(
  {
    1: {
      1: "She wished herself a happy birthday.",
    },
    2: {
      2: "He wished himself a happy birthday.",
    },
    "*": {
      "*": "They wished themselves a happy birthday.",
    },
  },
  [
    fbt._pronoun(3, gender, {
      human: 1,
    }),
    fbt._pronoun(2, gender, {
      human: 1,
    }),
  ],
  {
    hk: "3hCKKE",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle template descriptions 1`] = `
import { fbt } from "fbtee";
fbt._("Here it is", null, {
  hk: "124XBa",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should handle variations 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  {
    "*": "Click to see {count} links",
  },
  [fbt._param("count", c, [0])],
  {
    hk: "35KbmT",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should ignore __private attributes 1`] = `
import { fbt } from "fbtee";
fbt._("{foo}", [fbt._param("foo", foo)], {
  hk: "2kCEE7",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should ignore non-expression children in fbt:param 1`] = `
import { fbt } from "fbtee";
fbt._("{foo}", [fbt._param("foo", foo)], {
  hk: "2kCEE7",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should insert param value for same-param 1`] = `
import { fbt } from "fbtee";
fbt._("str {foo} and {foo}", [fbt._param("foo", Bar)], {
  hk: "4oznc3",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should maintain order of params and enums 1`] = `
import { fbt } from "fbtee";
fbt._(
  {
    x: {
      "*": "Hello, {foo}x{bar}",
    },
    y: {
      "*": "Hello, {foo}y{bar}",
    },
  },
  [
    fbt._enum(x, {
      x: "x",
      y: "y",
    }),
    fbt._param("bar", bar, [0, n]),
    fbt._param("foo", foo),
  ],
  {
    hk: "jYMse",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should not insert extra space 1`] = `
import { fbt } from "fbtee";
fbt._("Hello, {guest}!", [fbt._param("guest", guest)], {
  hk: "21cLqh",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should strip out more newlines 1`] = `
import { fbt } from "fbtee";
var x = fbt._("A simple string... with some other stuff.", null, {
  hk: "xpi1d",
});
baz();

`;

exports[`Test declarative (jsx) fbt syntax translation should strip out newlines 1`] = `
import { fbt } from "fbtee";
var x = fbt._("Preamble {parm}", [fbt._param("parm", blah)], {
  hk: "42eUy1",
});
baz();

`;

exports[`Test declarative (jsx) fbt syntax translation should support html escapes 1`] = `
import { fbt } from "fbtee";
fbt._("Ã—", null, {
  hk: "2jQoIS",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should support non-breasking space character 1`] = `
import { fbt } from "fbtee";
fbt._("text with non-breakingÂ Â Â space", null, {
  hk: "1WqZ9A",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should support unicode characters 1`] = `
import { fbt } from "fbtee";
// A backslash \\ in comments
fbt._(
  "A copyright sign Â©, a multi byte character ðŸ’©, and a backslash \\\\.",
  null,
  {
    hk: "4w0iJW",
  }
);

`;

exports[`Test declarative (jsx) fbt syntax translation should treat multiline descs as a single line 1`] = `
import { fbt } from "fbtee";
fbt._("lol", null, {
  hk: "1kGA01",
});

`;

exports[`Test declarative (jsx) fbt syntax translation should work with fragments 1`] = `
import { fbt } from "fbtee";
<Fragment>
  {fbt._(
    "{time} by {user name}",
    [
      fbt._param("time", formatDate(date, "F d, Y")),
      fbt._param(
        "user name",
        <Link
          href={{
            url: user.link,
          }}
        >
          {user.name}
        </Link>
      ),
    ],
    {
      hk: "1KGso4",
    }
  )}
</Fragment>;

`;

exports[`Test declarative (jsx) fbt syntax translation should work with implicit fragments 1`] = `
import { fbt } from "fbtee";
<>
  {fbt._(
    "{time} by {user name}",
    [
      fbt._param("time", formatDate(date, "F d, Y")),
      fbt._param(
        "user name",
        <Link
          href={{
            url: user.link,
          }}
        >
          {user.name}
        </Link>
      ),
    ],
    {
      hk: "1KGso4",
    }
  )}
</>;

`;

exports[`Test fbt transforms without the jsx transform [legacy buggy behavior] <fbt:pronoun> should insert a space character between two fbt constructs that don't neighbor raw text 1`] = `
import { fbt } from "fbtee";
fbt._(
  {
    "*": {
      "*": "You can add thesetags to anything.",
    },
    _1: {
      _1: "You can add this tag to anything.",
    },
  },
  [fbt._plural(count), fbt._plural(count)],
  {
    hk: "1kDgt0",
  }
);

`;

exports[`Test fbt transforms without the jsx transform nested in div 1`] = `
import { fbt } from "fbtee";
let x = (
  <div>
    {fbt._("A nested string", null, {
      hk: "2Z1yyy",
    })}
  </div>
);

`;

exports[`Test fbt transforms without the jsx transform not nested 1`] = `
import { fbt } from "fbtee";
let x = fbt._("A nested string", null, {
  hk: "2Z1yyy",
});

`;

exports[`Test fbt transforms without the jsx transform short bool syntax for doNotExtract attribute 1`] = `
import { fbt } from "fbtee";
let x = fbt._("Test", null, {
  hk: "2XNFBa",
});

`;

exports[`Test fbt transforms without the jsx transform short bool syntax for number attribute 1`] = `
import { fbt } from "fbtee";
let x = fbt._(
  {
    "*": "{name}",
  },
  [fbt._param("name", "name", [0])],
  {
    hk: "1SNt5t",
  }
);

`;

exports[`Test fbt transforms without the jsx transform should fail to preserve whitespace in text when preserveWhitespace=true (known bug) 1`] = `
import { fbt } from "fbtee";
fbt._("Some text with 3 spaces in between.", null, {
  hk: "3Uwc3D",
});

`;

exports[`Test fbt transforms without the jsx transform when using within template literals should dedupe plurals 1`] = `
import { fbt } from "fbtee";
fbt._(
  {
    "*": {
      "*": "There are {number} photos.",
    },
    _1: {
      _1: "There is 1 photo.",
    },
  },
  [fbt._plural(num), fbt._plural(num, "number", hi())],
  {
    hk: "2DWyfC",
  }
);

`;

exports[`Test fbt transforms without the jsx transform when using within template literals should work with a basic <fbt> 1`] = `
import { fbt } from "fbtee";
html\`<div>
  \${fbt._("basic text", null, {
    hk: "10IXnM",
  })}
</div>\`;

`;

exports[`Test fbt transforms without the jsx transform when using within template literals should work with a nested <fbt> within an <fbt:param> 1`] = `
import { fbt } from "fbtee";
html\`<div>
  \${fbt._(
    "outer text {param text}",
    [
      fbt._param(
        "param text",
        html\`<strong>
          \${fbt._(
            "inner text {inner param}",
            [fbt._param("inner param", "bold")],
            {
              hk: "koRSf",
            }
          )}
        </strong>\`
      ),
    ],
    {
      hk: "3keHFG",
    }
  )}
</div>\`;

`;

exports[`Test fbt transforms without the jsx transform when using within template literals should work with basic <fbt> auto-parameterization 1`] = `
import { fbt } from "fbtee";
html\`<div>
  \${fbt._(
    "outer text {=m1}",
    [
      fbt._implicitParam(
        "=m1",
        <strong>
          {fbt._("bold text", null, {
            hk: "3GKUYj",
          })}
        </strong>
      ),
    ],
    {
      hk: "4BN0aO",
    }
  )}
</div>\`;

`;
